miniml/compiler/compile.scm

```diff
@@ -2233,6 +2252,48 @@
      (compile-letexits env stacksize exit-defs body))
     (('LExit exit args)
      (compile-exit env stacksize exit args))
+    (('LWhile cond body)
+     (let* ((labcond (newlabel))
+            (labloop (newlabel)))
+       (bytecode-BRANCH-to labcond)
+       (bytecode-emit-label labloop)
+       (bytecode-put-u32-le CHECKSIGNALS)
+       (compile-expr env stacksize body)
+       (bytecode-emit-label labcond)
+       (compile-expr env stacksize cond)
+       (bytecode-put-u32-le BRANCHIF)
+       (bytecode-emit-labref labloop)
+       (bytecode-CONSTINT 0)
+       ))
+    (('LFor v dir b1 b2 body)
+     (let* ((labloop (newlabel))
+            (labend (newlabel)))
+       (compile-expr env stacksize b1)
+       (bytecode-put-u32-le PUSH)
+       (compile-expr env (+ stacksize 1) b2)
+       (bytecode-put-u32-le PUSH)
+       (bytecode-put-u32-le PUSH)
+       (bytecode-ACC 2)
+       (bytecode-put-u32-le (if (equal? dir 'UpTo) GTINT LTINT))
+       (bytecode-put-u32-le BRANCHIF)
+       (bytecode-emit-labref labend)
+       (bytecode-emit-label labloop)
+       (bytecode-put-u32-le CHECKSIGNALS)
+       (compile-expr (stack-var env v stacksize) (+ stacksize 2) body)
+       (bytecode-ACC 1)
+       (bytecode-put-u32-le PUSH)
+       (bytecode-put-u32-le OFFSETINT)
+       (bytecode-put-u32-le (if (equal? dir 'UpTo) 1 -1))
+       (bytecode-put-u32-le ASSIGN)
+       (bytecode-put-u32-le 2)
+       (bytecode-ACC 1)
+       (bytecode-put-u32-le NEQ)
+       (bytecode-put-u32-le BRANCHIF)
```

Comment:
Here is my informal notation for stack+accumulator tracking (sometimes I wonder if we should add such things as comments to understand this sort of code):

```
labloop:
CHECKSIGNALS
<body>            ; b1 b2 | body
ACC 1             ; b1 b2 | b1
PUSH              ; b1 b2 b1 | b1
OFFSETINT         ; b1 b2 b1 | b1+1 (for UPTO)
ASSIGN 2          ; b1+1 b2 b1 | b1+1
ACC 1             ; b1+1 b2 b1 | b2
NEQ               ; b1+1 b2 | b1 != b2
BRANCHIF labloop
```

This is very nitpicky, but I believe we could also use the following sequence, which could be one opcode less in a future world with on-the-fley PUSHACC1 compression (or if we just defined and used it statically):

```
labloop:
CHECKSIGNALS
<body>            ; b1 b2 | body
ACC 1             ; b1 b2 | b1
OFFSETINT         ; b1 b2 | b1+1 (for UPTO)
ASSIGN 1          ; b1+1 b2 | b1+1
PUSH              ; b1+1 b2 b1+1 | b1+1
ACC 1             ; b1+1 b2 b1+1 | b2
LEINT             ; b1+1 b2 | b1+1 <= b2
BRANCHIF labloop
```

Reply:
For the record, here is the version where `b2` is a variable:

```
LFor v dir b1 x2 body

<b1>                       ; | b1
PUSH                       ; b1 | b1
PUSH                       ; b1 b1 | b1
<x2>                       ; b1 b1 | x2
(dir=upto ? LTINT : GTINT) ; b1 | (x2 < b1) (for UPTO)
BRANCHIF labend

labloop:          ; b1 | _
CHECKSIGNALS
<body>            ; b1 | body
ACC 0             ; b1 | b1
OFFSETINT         ; b1 | b1+1 (for UPTO)
ASSIGN 0          ; b1+1 | b1+1
PUSH              ; b1+1 b1+1 | b1+1
<x2>              ; b1+1 b1+1 | x2
GEINT             ; b1+1 | x2 >= b1+1
BRANCHIF labloop

labend:    ; b1 | _
CONSTINT 0 ; b1 | 0
POP 1      ; | 0
```

It is exactly as long, but I find it slightly simpler to tell what is going on. This being said, now that you have written the two-stack version and that I have carefully reviewed it, I don't think there is any point in changing.

Reply:
True indeed. I'll merge it as is, but I'll keep the idea in mind for future features, if needed.